apple/swift-docc
# Highlight declaration differences in overloaded symbol groups

Bug/issue #, if applicable: rdar://116409531

## Summary

This PR introduces a new field to declaration tokens in Render JSON: `highlight`. This field indicates that the token is distinct among a set of overloads and should be indicated with a highlight or other style. It also performs a difference comparison of overloaded symbols' declarations so that the fields are populated for Swift-DocC-Render.

When combined with the related Swift-DocC-Render PR (linked below), the rendered page appears like this:

<img width="916" alt="image" src="https://github.com/apple/swift-docc/assets/5217170/c48d1cf2-fe7e-45ef-936d-a307bcdc3e82">

Because the differencing works directly on the declaration fragments received from the symbol graph, this PR also takes some care to break up `.text` tokens to prevent false-positive differences and create a cleaner output. For example, the below overloads correctly only highlight the `<S>` in the second overload, even though it originally contains a text token of `>(`. Also visible in this example is some proactive whitespace-trimming from highlighted sections; there is a unique text token containing a single space before the `where` clause in the second overload, but it's not highlighted to reduce clutter:

<img width="913" alt="image" src="https://github.com/apple/swift-docc/assets/5217170/4883dced-5c73-4a6e-9246-a22b54b989d7">

### Performance

This PR uses the standard library's `CollectionDifference` API to generate the [Longest Common Subsequence](https://en.wikipedia.org/wiki/Longest_common_subsequence) of the overload declarations. The diff algorithm can have a heavy time complexity, especially when run repeatedly like here. The standard library's implementation is well-optimized, but there is still a necessary performance hit to calculate the diffs. This following is a benchmark run on a large framework with many overloaded symbols:

```
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Metric                                   │ Change          │ main                 │ current              │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ Duration for 'bundle-registration'       │ +1.913%¹,²      │ 10.351 sec           │ 10.549 sec           │
│ Duration for 'convert-total-time'        │ +3.857%³,⁴      │ 15.477 sec           │ 16.074 sec           │
│ Duration for 'documentation-processing'  │ +7.959%⁵,⁶      │ 5.003 sec            │ 5.401 sec            │
│ Duration for 'finalize-navigation-index' │ no change⁷      │ 0.046 sec            │ 0.046 sec            │
│ Peak memory footprint                    │ no change⁸      │ 811 MB               │ 835.4 MB             │
│ Data subdirectory size                   │ +2.028%⁹        │ 127.4 MB             │ 130 MB               │
│ Index subdirectory size                  │ -0.02829%¹⁰     │ 1.1 MB               │ 1.1 MB               │
│ Total DocC archive size                  │ +1.092%¹¹       │ 236.6 MB             │ 239.2 MB             │
│ Topic Anchor Checksum                    │ no change       │ 53074b40863239b7a00d │ 53074b40863239b7a00d │
│ Topic Graph Checksum                     │ change          │ 5bdfe323b77baec3d5cd │ f4af046ddef9eab49ba8 │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

## Dependencies

https://github.com/apple/swift-docc-render/pull/847 is required to render the tokens with a highlight.

## Testing

Steps:
1. Build the Swift-DocC-Render PR and link the resulting `dist` directory into the `DOCC_HTML_DIR` environment variable.
2. `swift run docc preview --enable-experimental-overloaded-symbol-presentation 'Tests/SwiftDocCTests/Test Bundles/OverloadedSymbols.docc'
3. Navigate to the `OverloadedEnum/firstTestMemberName(_:)` overload group and expand the declaration list.
4. Ensure that the declarations correctly highlight the parameter type in each overloaded declaration.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [ n/a ] Updated documentation if necessary
